File: Assignment_logreg_affairs.py
# Logistic Regression - affairs dataset

# Business Problem: To classify persons having affairs and those not having any affairs


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression

# EDA
affairs_raw = pd.read_csv("D:\\affairs.csv")
affairs_raw.columns
affairs_raw.head()
affairs_raw.describe()
affairs_raw.dtypes

affairsdata = affairs_raw.copy()
affairsdata['affairs']=(affairsdata.affairs >0).astype(int)
affairsdata.affairs.value_counts()
# 0 451 and  1 150

affairsdata.groupby('affairs').mean()
''' there is very less difference in variables in terms of mean. those who are not
having affairs have given higher rating for marriage (more happier), former are
more religious
'''

# visualizations
sns.countplot(x="affairs",data=affairsdata)
# the proportion of thos not having affairs is more than that of having affairs
sns.countplot(x="gender",data=affairsdata)
# our dataset has a few more females than males
sns.countplot(x="children",data=affairsdata)
# lesser persons in the dataset have no children as compared to those havng children
affairsdata.religiousness.hist()

# barplot of religiousness grouped by affair
pd.crosstab(affairsdata.religiousness, affairsdata.affairs).plot(kind='bar')
# barplot of religiousness grouped by affair
pd.crosstab(affairsdata.religiousness, affairsdata.affairs).plot(kind='bar')
# barplot of yearsmarried grouped by affair
pd.crosstab(affairsdata.yearsmarried, affairsdata.affairs).plot(kind='bar')

# finding missing values in the data 

affairsdata.isnull().sum() # no missing values

# creating dummy variables for categorical varaibles
affairs_dummies = pd.get_dummies(affairsdata[["gender","children"]])
# Dropping the columns for which we have created dummies
affairsdata.drop(["gender","children"],inplace=True,axis = 1)

# adding the columns to the affairsdata data frame 

affairsdata = pd.concat([affairsdata,affairs_dummies],axis=1)
##################### Model building

affairsdata.shape # 601 records ,11 variables
X = affairsdata.iloc[:,[1,2,3,4,5,6,7,8,9,10]]
Y = affairsdata.iloc[:,0]
classifier = LogisticRegression()
classifier.fit(X,Y)

classifier.coef_ # coefficients of features 
classifier.predict_proba (X) # Probability values 

y_pred = classifier.predict(X)
affairsdata["y_pred"] = y_pred
y_prob = pd.DataFrame(classifier.predict_proba(X.iloc[:,:]))
new_df = pd.concat([affairsdata,y_prob],axis=1)
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(Y,y_pred)
print (confusion_matrix)

# accuracy 
from sklearn import metrics
classifier.score(X,Y) # 76%
(437+20)/601
affairsdata.affairs.value_counts()
'''
here actuals are in the rows,  though accuracy is 76% , True postive is very less 20
our of 150.'''

# evaluate model using 10-fold cross validatation
from sklearn.model_selection import cross_val_score
scores = cross_val_score(LogisticRegression(),X,Y,scoring='accuracy', cv=10)
print(scores)
# [0.72131148 0.73333333 0.73333333 0.75       0.76666667 0.75
# 0.76666667 0.78333333 0.73333333 0.76666667]
scores.mean() # 75%

#the accuracy is same as when we used the complete dataset to train and predict 

'''
CONCLUSIONS

we are classifying persons those who are having affairs and those who are not
having affairs. We are using logistic regression to do the binary calssification.

we have built a standard model using complete dataset. We have evaluated model
by predicting on the whole dataset and also used 10-fold cross validation.

The accuracies obtained were same in both. We can improve the accuracy by using other
classification techniques like Naive Bayes, KNN and also by increasing the sample size.
'''

File: Assignment_Logreg_affairs.R
# Logistic Regression Assignment on affairs dataset

# Business Problem: To classify persons having and not having affairs

# Data collection: 
affairs_raw <- read.csv("D:\\Logistic_Regression_Assignment\\affairs.csv")
# 601 records of 9 variables

# some info about the dataset
'''
The details of variables in this dataset is given below:
Gender Identification (male, female)
Age
Number of Years Married
Children (yes, no)
How Religious, from 1 (not religious) to 5 (very religious)
Education
Occupation, from 1 to 7, according to hollingshead classification
Self Rating of Marriage, from 1 (very unhappy) to 5 (very happy)
Number of Affairs in the Past Year
'''
'''
since affairs is a binary variable (person has affairs or does not have), we can
fit binary Logistic Regression to classify the data
'''

head(affairs_raw)
summary(affairs_raw)
str(affairs_raw)
names(affairs_raw)
#  "affairs"       "gender"        "age"           "yearsmarried"  "children"     
#  "religiousness" "education"     "occupation"    "rating" 

attach(affairs_raw)
table(affairs)
table(gender)
table(children)


# affairs is the outcome variable and a person has affairs or not 
# if he/she has no affairs using 0 and has affairs using 1 to represent the same.
affairs_raw1 <- affairs_raw
affairs_raw1$affairs <- ifelse(affairs_raw1$affairs>0,1,0)
# also class (data type) of affairs is given as int, let us convert it to factor
class(affairs_raw1$affairs)
affairs_raw1$affairs <- as.factor(affairs_raw1$affairs)

table(affairs_raw1$affairs)
#   0   1 
#  451 150

prop.table(table(affairs_raw1$affairs))
#        0        1 
#  0.750416 0.249584 
# there are 75% persons not having affairs and 25% who have affairs

# there are 2 categorical variables. to know how R dummifies them 
contrasts(affairs_raw1$gender)
contrasts(affairs_raw1$children)
# in gender, female is 0, it will be used as reference
# in children, no is 0, it will be used as reference.

# checking for missing values
sum(is.na(affairs_raw1))
library(Amelia)
missmap(affairs_raw1,main='Observed vs missing values')
# the whole plot is blue (for observed values), no missing values

# Building Logistic Regression model
logreg1 <- glm(affairs ~ ., data=affairs_raw1,family='binomial')

summary(logreg1)
# null deviance is 675, Resid dev is 609 and AIC is 627
'''
in logit model, response variable is given as natural log of odds. When rating 
increases by 1 unit, log odds reduces by 0.47. 
'''
# To calculate the odds ratio manually 
exp(coef(logreg1))


# analyze the table of deviance uisng anova
anova(logreg1, test='Chisq')
'''
the difference in null and resid. dev shows how our model is doing against
the null model. greater the difference, better is our model.
from table we can see drop in deviancce when each variable is added one at
a time. 
'''

# gender, children, education and occupation are not significant
# let us remove these and build another model

logreg2 <- glm(affairs ~ age+yearsmarried+education+occupation, data=affairs_raw1,family='binomial')

summary(logreg2)
# null deviance is 675, Resid dev is 657 and AIC is 667
# AIC is more and even resid. deviance is more as compared to previous model.
# so logreg1 is a better model

# let us predict the values, we get the probabilities 
pred_prob <- logreg1$fitted.values
head(logreg1$fitted.values)
# pred_prob=predict(logreg1,type=c("response"),affairs_raw1)

# let us now look at the accuracy measures
# confusion matrix
confusion <- table(pred_prob>=0.5, affairs_raw1$affairs)
confusion
table(affairs_raw1$affairs)

Accuracy<-sum(diag(confusion))/sum(confusion)
Accuracy # 75%
# but we can see that the model has correctly predicted No as NO.
# all Yes are wrongly predicted by the model.
# here alpha error (false positive) is zero, which is very good.

# ROC Curve 
library(ROCR)
rocrpred<-prediction(pred_prob, affairs_raw1$affairs)
rocrperf<-performance(rocrpred,'tpr','fpr')
plot(rocrperf,colorize=T)
# More area under the ROC Curve better is the logistic regression model obtained

# auc
auc <- performance(rocrpred,measure='auc')
auc <- auc@y.values[[1]]
auc # 0.61
# the auc is 61%. area should be cclose to 1 and far from 0.5, to be a good
# model. 

#########################   Build model using train 
library(caTools)
set.seed(100)
split_data <- sample.split(affairs_raw1$affairs, SplitRatio=0.8)
train <- subset(affairs_raw1, split_data==TRUE)
test <- subset(affairs_raw1, split_data==FALSE)

logreg3 <- glm(affairs ~ ., data=train,family='binomial')
summary(logreg3)
# null dev is 540, resid dev is 482 and AIC is 500 which is far less than 
#previous models, hence a better model.

# training accuracy
pred_prob3 <- logreg3$fitted.values
# confusion matrix
confusion3 <- table(pred_prob3>=0.5, train$affairs)
confusion3
table(train$affairs)
train_Accuracy<-sum(diag(confusion3))/sum(confusion3)
train_Accuracy # 76.7%

# testing accuracy
pred_test <- predict(logreg3,test,type='response')
pred_test[1:5]
confusion_test <- table(pred_test>=0.5, test$affairs)
confusion_test
table(test$affairs)
test_Accuracy<-sum(diag(confusion_test))/sum(confusion_test)
test_Accuracy # 75.8%
# both train and test accuracy are almost same. 

# ROCR
rocrpred3<-prediction(pred_prob3, train$affairs)
rocrperf3<-performance(rocrpred3,'tpr','fpr')
plot(rocrperf3,colorize=T)
abline(a=0, b=1, col='black')
# auc
auc3 <- performance(rocrpred3,measure='auc')
auc3 <- auc3@y.values[[1]]
auc3 # 72.4%

# auc for test data
rocrpred3_test <- prediction(pred_test,test$affairs)
auc3_test <- performance(rocrpred3_test, measure='auc')
auc3_test <- auc3_test@y.values[[1]]
auc3_test # 65.85%

# the auc for previous model (without splitting the data) was 61%. After splitting
# the data it has increases for both train data and test data.

#let us compute optimal cutoff value
library(InformationValue)
opt_cutoff = optimalCutoff(test$affairs,pred_test)[1]
opt_cutoff # 0.5388419

confusion_test1 <- table(pred_test>=0.54, test$affairs)
confusion_test1
table(test$affairs)
test_Accuracy1<-sum(diag(confusion_test1))/sum(confusion_test1)
test_Accuracy1 # 77.5%

# for cut_off 0.5, test-accuracy is 75.8%
# for cut_off 0.5388, test-accuracy is 77.5%
# for cut_off 0.54, test-accuracy is 77.5%
# for cut_off 0.6, test-accuracy is 75%

'''
the accuracy is decreasing bothways if we move from the optimal cutoff value.
The false positive has decreased when we consider cuoff value.
'''
'''
CONCLUSIONS

The business problem is to classify persons having affairs from others not having
affairs. Since the output variable is categorical we have used logistic 
regression technique.

We have prepared various models, and chosen the better model based on
accuracy, ROC and auc.  we have also used optimal cut-off value to reduce
misclassification errors.

To increase our accuracy and decrease misclassification, we need more records.

'''



